lib/logstash/inputs/jms.rb

```diff
@@ -180,102 +296,54 @@ def queue_event(msg, output_queue)
-      connection.on_message(destination_key => @destination, :selector => @selector) do |message|
-        queue_event message, output_queue
-      end
-      # Since the on_message handler above is in a separate thread the thread needs
-      # to do some other work. It will just sleep for 10 seconds.
-      while !stop?
-        @logger.debug("JMS Thread sleeping ...")
-        sleep @interval
-      end
-    end
-  rescue => e
-    @logger.warn("JMS Consumer died", :exception => e, :backtrace => e.backtrace)
-    sleep(10)
-    retry unless stop?
-  end # def run_async
+    params[destination_key] = @destination
+    queue_or_topic = session.create_destination(params)
+    @durable_subscriber ? durable_subscriber(session, queue_or_topic, params) :
+                          regular_subscriber(session, queue_or_topic, params)
+  end
 
-  public
-  def run(output_queue)
-    case @runner
-    when "consumer" then
-      run_consumer(output_queue)
-    when "async" then
-      run_async(output_queue)
-    when "thread" then
-      run_thread(output_queue)
-    end
-  end # def run
 
-  public
-  def close
-    @logger.info("Closing JMS connection")
-    @connection.close rescue nil
-  end # def close
+  def durable_subscriber(session, queue_or_topic, params)
+    params[:selector]  ? session.create_durable_subscriber(queue_or_topic, @durable_subscriber_name, params[:selector], false) :
+                         session.create_durable_subscriber(queue_or_topic, @durable_subscriber_name)
+  end
 
-  public
-  def stop
-    @logger.info("Stopping JMS consumer")
-    @connection.stop rescue nil
-  end # def stop
+  def regular_subscriber(session, queue_or_topic, params)
+    params[:selector] ? session.create_consumer(queue_or_topic, params[:selector]) :
+                        session.create_consumer(queue_or_topic)
+  end
+
+  def error_hash(e)
+    error_hash = {:exception => e.class.name, :exception_message => e.message, :backtrace => e.backtrace}
+    root_cause = get_root_cause(e)
+    error_hash[:root_cause] = root_cause unless root_cause.nil?
+    error_hash
+  end
+
+  # JMS Exceptions can contain chains of Exceptions, making it difficult to determine the root cause of an error
+  # without knowing the actual root cause behind the problem.
+  # This method protects against Java Exceptions where the cause methods loop. If there is a cause loop, the last
+  # cause exception before the loop is detected will be returned, along with an entry in the root_cause hash indicating
+  # that an exception loop was detected.
+  def get_root_cause(e)
+    return nil unless e.respond_to?(:get_cause) && !e.get_cause.nil?
+    cause = e
+    slow_pointer = e
+    # Use a slow pointer to avoid cause loops in Java Exceptions
+    move_slow = false
+    until (next_cause = cause.get_cause).nil?
+      cause = next_cause
+      return {:exception => cause.class.name, :exception_message => cause.message, :exception_loop => true } if cause == slow_pointer
```

Comment:
Other than being useful in testing, what can the user do with this info `:exception_loop => true`? Is there guidance? Should there be?

Reply:
The intention was to make users/us aware when troubleshooting that exception chain may not be complete, which I thought may be helpful info

Reply:
It is not something I really expect to see in reality, but it is a possibility that it could occur in code that is outside of our control

Reply:
Ahh. Don't get me wrong, I'm not knocking it. In the light of your reply then maybe a note in the docs to say that if an exception loop is detected when finding the root cause of a JMS Exception then the reported root cause might or might not be accurate.

Reply:
I've added a little more detail to the comment. Is that what you mean? We typically don't mention exceptions/errors in our official documentation to the best of my knowledge

Reply:
The addition is fine.